server/src/main/java/org/apache/cassandra/sidecar/concurrent/AsyncConcurrentTaskExecutor.java

```diff
@@ -0,0 +1,285 @@
+ *
+ * <h2>Architecture</h2>
+ * <p>The executor uses a promise-based triggering mechanism:</p>
+ * <ul>
+ *   <li>Each task is associated with a {@link Promise} that acts as a trigger signal</li>
+ *   <li>Tasks only start execution when their corresponding Promise completes successfully</li>
+ *   <li>Initially, {@code maxConcurrency} number of promises are completed to start the execution flow</li>
+ *   <li>As tasks complete, the next pending task is automatically triggered</li>
+ * </ul>
+ *
+ * <h2>Concurrency Control</h2>
+ * <p>The executor maintains a global task index using {@link AtomicInteger} to ensure thread-safe
+ * task scheduling. Tasks are processed in the order they were submitted, with automatic
+ * flow control to maintain the desired concurrency level.</p>
+ *
+ * <h2>Error Handling</h2>
+ * <p>When a task completes (successfully or with a non-{@link CancellationException} error),
+ * the executor automatically triggers the next pending task. Tasks that are cancelled do not
+ * trigger subsequent tasks to prevent further executions.</p>
+ *
+ * <h2>Usage Example</h2>
+ * <pre>{@code
+ * List<Callable<Future<String>>> tasks = Arrays.asList(
+ *     () -> downloadFile("file1.txt"),
+ *     () -> downloadFile("file2.txt"),
+ *     () -> downloadFile("file3.txt")
+ * );
+ *
+ * AsyncConcurrentTaskExecutor<String> executor =
+ *     new AsyncConcurrentTaskExecutor<>(vertx, tasks, 2); // max 2 tasks can execute concurrently
+ *
+ * List<Future<String>> results = executor.start();
+ * }</pre>
+ *
+ * @param <T> the type of result produced by the tasks
+ */
+public class AsyncConcurrentTaskExecutor<T>
+{
+    protected static final String TASK_CANCEL_MESSAGE = "task cancelled";
+    private final AtomicInteger index = new AtomicInteger(0);
+    private final int maxConcurrency;
+    private final Vertx vertx;
+    private final List<Promise<T>> taskPromises;
+    private final List<Callable<Future<T>>> tasks;
+
+    /**
+     * Creates a new AsyncConcurrentTaskExecutor.
+     *
+     * @param vertx          the Vert.x instance for context execution
+     * @param tasks          the list of tasks to execute (must not be empty)
+     * @param maxConcurrency the maximum number of tasks to execute concurrently (must be &gt; 0)
+     * @throws IllegalArgumentException if maxConcurrency &lt;= 0 or tasks is empty
+     */
+    public AsyncConcurrentTaskExecutor(@NotNull Vertx vertx,
+                                       @NotNull List<Callable<Future<T>>> tasks,
+                                       int maxConcurrency)
+    {
+        Objects.requireNonNull(vertx, "vertx must not be null.");
+        Objects.requireNonNull(tasks, "tasks must not be null.");
+        if (maxConcurrency < 1)
+        {
+            throw (new IllegalArgumentException("maxConcurrency must be > 0"));
+        }
+
+        this.tasks = Collections.unmodifiableList(tasks);
```

Comment:
NIT: prefer newer JDK APIs
```suggestion
        this.tasks = List.copyOf(tasks);
```

Reply:
done